### PR TITLE
add linting on notebooks with nbqa

### PR DESCRIPTION
**Pull request by jameslamb**
I've been moving a lot of things around in the notebooks, and found myself making some mistakes. For example, I think that in #59 , I accidentally broke one of the XGBoost notebooks because a piece of code was relying on an `import os` from a line that was higher up.

I've also found that it's easy to leave unnecessary imports behind (#29, #44, #51).

I've also found that there some review time is spent on small style things like trailing whitespace.

This PR proposes fixes for these problems by introducing linting on the notebooks in this project, via [nbqa](https://github.com/nbQA-dev/nbQA). As of this PR, all of the Python code in notebooks would have to pass `flake8` and `black`.

### Changes in this PR

This PR proposes a fix for that:

1. `make format` now applies [`black`](https://github.com/psf/black) formatting to all Python code in notebooks
2. `make lint` now runs `black` over all Python code in notebooks, and fails builds if any issues are found
3. `make lint` now runs `flake8` over all Python code in notebooks, and fails builds if any issues are found

### How this improves `examples`

I think that this change will help us to automatically catch issues before they're merged to `main`, which is preferable to catching them in integration tests (expensive) or in user-facing issues that lead to support questions (:cry: ).

These checks will catch some legitimate code-correctness things (like unused variables and imports) and will enforce a consistent style so that we don't spend reviewer or contributor time and energy talking about style things like trailing whitespace or single quotes vs. double quotes.

See the list of things caught by `flake8`, for example, and fixed in this PR

<details><summary>flake8 issues</summary>

```text
examples/examples-cpu/snowflake/snowflake-dask.ipynb:cell_20:1:1: F811 redefinition of unused 'wait' from line 44
examples/examples-cpu/nyc-taxi/data-aggregation.ipynb:cell_2:5:1: F401 'hvplot.pandas' imported but unused
examples/examples-cpu/nyc-taxi/data-aggregation.ipynb:cell_2:5:19: E401 multiple imports on one line
examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb:cell_2:1:1: F401 'dask.distributed.wait' imported but unused
examples/examples-cpu/nyc-taxi/dashboard.ipynb:cell_1:6:1: F401 'hvplot.pandas' imported but unused
examples/examples-cpu/nyc-taxi/dashboard.ipynb:cell_1:6:19: E401 multiple imports on one line
examples/examples-cpu/nyc-taxi/dashboard.ipynb:cell_1:7:1: F401 'holoviews as hv' imported but unused
examples/examples-cpu/nyc-taxi/dashboard.ipynb:cell_16:23:11: F541 f-string is missing placeholders
examples/examples-cpu/nyc-taxi-snowflake/hyperparameter-scikit.ipynb:cell_2:3:1: F401 'datetime' imported but unused
examples/examples-cpu/nyc-taxi-snowflake/hyperparameter-scikit.ipynb:cell_2:4:1: F401 'pandas as pd' imported but unused
examples/examples-cpu/nyc-taxi-snowflake/hyperparameter-scikit.ipynb:cell_2:5:1: F401 's3fs' imported but unused
examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:cell_1:5:1: F401 'hvplot.pandas' imported but unused
examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:cell_1:5:19: E401 multiple imports on one line
examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:cell_1:6:1: F401 'holoviews as hv' imported but unused
examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:cell_4:5:28: F541 f-string is missing placeholders
examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:cell_28:23:11: F541 f-string is missing placeholders
examples/examples-cpu/nyc-taxi-snowflake/xgboost.ipynb:cell_2:3:21: F821 undefined name 'os'
examples/examples-cpu/nyc-taxi-snowflake/xgboost.ipynb:cell_2:4:18: F821 undefined name 'os'
examples/examples-cpu/nyc-taxi-snowflake/xgboost.ipynb:cell_2:5:22: F821 undefined name 'os'
examples/examples-cpu/nyc-taxi-snowflake/xgboost.ipynb:cell_2:7:23: F821 undefined name 'os'
examples/examples-cpu/nyc-taxi-snowflake/xgboost.ipynb:cell_2:8:17: F821 undefined name 'os'
examples/examples-cpu/nyc-taxi-snowflake/xgboost.ipynb:cell_2:9:15: F821 undefined name 'os'
examples/examples-cpu/prefect/prefect-cloud-scheduled-scoring.ipynb:cell_1:11:1: F401 'prefect.Task' imported but unused
examples/examples-cpu/prefect/prefect-scoring.ipynb:cell_1:8:1: F401 'datetime.timedelta' imported but unused
examples/examples-cpu/prefect/prefect-scoring.ipynb:cell_1:10:1: F401 'prefect.Task' imported but unused
examples/examples-gpu/nyc-taxi-snowflake/rf-rapids-dask.ipynb:cell_5:3:1: F401 'dask.dataframe as dd' imported but unused
```

</details>

As a result, I think this change will allow us to move more quickly, with greater confidence in releases.

### Notes for reviewers

This is a potentially disruptive change, so I'd like everyone added as reviewers to approve. If you disagree, please explain why here and we can try to come to some compromise.

I've adding @skirmer @rikturr @jsignell and @bhperry as reviewers because you all work on this repo and have a stake in what it takes to get a PR merged here.

Thanks for you time and consideration.

### CODE REVIEW

**Comment by jameslamb** [comment]
ok actually @jsignell ignore this PR for now. I just learned about https://github.com/nbQA-dev/nbQA which might be even better and not require us to maintain this weird script I wrote.

I'm going to move this back to draft.

**Comment by jameslamb** [comment]
Forgot to add the list of `flake8` issues caught by this new linting and fixed in this PR. I just added that to the PR description. I think it gives some good examples of what we'd gain by adding these checks.

File: examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb [hunk 1]
```diff
@@ -61,13 +61,12 @@
     "import datetime as dt\n",
     "import numpy as np\n",
     "\n",
-    "import hvplot.dask, hvplot.pandas\n",
```
Comment (jsignell) [inline examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:64]: These imports are needed they extend the dask.Dataframe and pandas.DataFrame classes to support `.hvplot`.

Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:64]: oh ok. I guess that would explain the `#noqa` in these examples :grinning: --> https://hvplot.holoviz.org/user_guide/Plotting.html#plotting

Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:64]: put them back: https://github.com/saturncloud/examples/pull/60/commits/4c721a4d24a58202c9c751622e6c95bc75a50d1a

**Review by jsignell — APPROVED** [review]
I skimmed it, but looks good to me. Did you see all the pre-commit hooks that the nbqa folks have :grimacing: ??? https://github.com/nbQA-dev/nbQA/blob/master/.pre-commit-config.yaml

**Comment by jameslamb** [comment]
> I skimmed it, but looks good to me. Did you see all the pre-commit hooks that the nbqa folks have  ??? https://github.com/nbQA-dev/nbQA/blob/master/.pre-commit-config.yaml

haha wow, that is a LOT

File: examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb [hunk 2]
```diff
@@ -61,13 +61,14 @@
     "import datetime as dt\n",
```
Comment (rikturr) [inline examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:61]: <p>this autoformat seems strange to me, I would prefer the old version</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/saturncloud/examples/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

File: examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb [hunk 3]
```diff
@@ -35,19 +35,19 @@
    "outputs": [],
```
Comment (rikturr) [inline examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb:35]: <p>autoformat here too, I think old way is better</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/saturncloud/examples/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

**Review by rikturr — COMMENTED** [review]
Looks good overall! I am lamenting black's preference for double quotes only because I single quote all my strings 😬 . I did find a couple places specifically in function calls where we wanted each parameter in a new line but then the autoformat put it back down to one line. I'm not sure if there's a way around this?

**Review by bhperry — APPROVED** [review]
LGTM

File: examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb (hunk 3 above)
Comment (jsignell) [reply examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb:35]: <p>probably a reasonable way around this is to define the kwargs above. Like </p><p> </p><pre class="ql-syntax" spellcheck="false">data = taxi_train[features]
label = taxi_train[y_col]
dtrain = xgb.dask.DaskDMatrix(client=client, data=data, label=label)
</pre><p> </p>

File: examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb (hunk 2 above)
Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:61]: <p>oy yeah I'd prefer the old way too. But the point of <code>black</code>  is that you give up caring about exactly what it looks like, and what you get is the freedom of never thinking about it again.</p><p> </p><p>I think we can trick this into giving us what we want by adding another header like <code>Accept: "application/json"</code>.  Honestly we should have had that anyway.</p>

File: examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb (hunk 3 above)
Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb:35]: <p>the relevant line is 96 characters, so if I just make the variable <code>dtrain_dmatrix</code>,  I think <code>black</code>  will  force a line break and give us back the formatting we want :)</p>

Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb:35]: <p>ha nope that didn't work. I like Julia's idea.</p>

Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi-snowflake/xgboost-dask.ipynb:35]: I did this in https://github.com/saturncloud/examples/pull/60/commits/6869031b618eb9c7709864185fab0ac0a6513534

File: examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb (hunk 2 above)
Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi-snowflake/dashboard.ipynb:61]: ok I think this had the desired effect! https://github.com/saturncloud/examples/pull/60/commits/6869031b618eb9c7709864185fab0ac0a6513534

image

**Comment by jameslamb** [comment]
> Looks good overall! I am lamenting black's preference for double quotes only because I single quote all my strings  .

I think the benefit (never having to think about it or have PR comments about it) is worth the cost of it not matching your preferences, especially for something like single-quotes vs double-quotes that doesn't impact the correctness of the code. The reality is that right now in the examples we have a mix of single-quote and double-quote, because we don't have any mechanism to keep it consistent.

And since `make format` will auto-format everything, you're welcome to write in any style you're used to and let the automation change things.

> I did find a couple places specifically in function calls where we wanted each parameter in a new line but then the autoformat put it back down to one line. I'm not sure if there's a way around this?

`black` is famously, intentionally, not configurable. You can use some tricks to get it to break things onto multiple lines, like preferring keyword arguments and giving variables longer names. For the places where we care a lot about the line breaks, we can invest a little effort into those tricks. I know it feels annoying and weird, but we use `black` on most of the other projects in Saturn and I think it's been a huge boost to my productivity (compared to projects I've worked on in the past that don't use it).

**Comment by jsignell** [comment]
Yeah I have to agree with James. I am a slow reader and having a consistent style really helps me get the meaning faster. I know it seems silly, but it's a real thing.

**Review by skirmer — APPROVED** [review]
Looks good to me- I read about half closely and skimmed the others. I know Aaron has feels about the single vs double quotes but I don't have an issue with that.

**Comment by jameslamb** [comment]
@rikturr are there any other questions I can answer for you on this PR? This touches a lot of files so I'd like to get it merged to avoid conflicts in the next batch of PRs.

**Review by rikturr — APPROVED** [review]

**Pull request merged**